File: csharp/BronKerbosch/BronKerbosch2bGP.cs
// Bron-Kerbosch algorithm with pivot of highest degree (IK_GP)

using BronKerbosch;
using System.Collections.Generic;

internal static class BronKerbosch2bGP<VertexSet, VertexSetMgr, TAccumulator>
    where VertexSet : ISet<Vertex>
    where VertexSetMgr : IVertexSetMgr<VertexSet>
    where TAccumulator : ICliqueAccumulator<TAccumulator>, new()
{
#   pragma warning disable IDE0022 // Use expression body for method
    public static void Explore(UndirectedGraph<VertexSet, VertexSetMgr> graph,
                               CliqueConsumer<TAccumulator> consumer)
    {
        Pivot<VertexSet, VertexSetMgr, TAccumulator>.Explore(graph, consumer,
                                                             PivotChoice.MaxDegreeLocal);
    }
}

File: go/BronKerbosch/bron_kerbosch2b_gp.go
package BronKerbosch

func bronKerbosch2bGP(graph *UndirectedGraph, consumer Consumer) {
	// Bron-Kerbosch algorithm with pivot of highest degree within remaining candidates
	// chosen from candidates only (IK_GP).
	order := graph.Order()
	if order > 0 {
		pivot := graph.maxDegreeVertex()
		// In this initial iteration, we don't need to represent the set of candidates
		// because all neighbours are candidates until excluded.
		excluded := make([]bool, order)
		for i := range order {
			v := Vertex(i)
			neighbours := graph.neighbours(v)
			if !neighbours.IsEmpty() && !neighbours.Contains(pivot) {
				neighbouringExcluded := neighbours.IntersectionWithMap(excluded)
				if len(neighbouringExcluded) < len(neighbours) {
					neighbouringCandidates := neighbours.Difference(neighbouringExcluded)
					visit(
						graph, consumer,
						MaxDegreeLocal,
						neighbouringCandidates,
						neighbouringExcluded,
						[]Vertex{v})
				}
				excluded[v] = true
			}
		}
	}
}
